src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/CompletionTriggerAndCommitCharacters.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.Completion;
+
+internal class CompletionTriggerAndCommitCharacters
+{
+    /// <summary>
+    ///  Trigger character that can trigger both Razor and Delegation completion
+    /// </summary>
+    private const char TransitionCharacter = '@';
+
+    private static readonly char[] s_vsHtmlTriggerCharacters = [':', '#', '.', '!', '*', ',', '(', '[', '-', '<', '&', '\\', '/', '\'', '"', '=', ':', ' ', '`'];
+    private static readonly char[] s_vsCodeHtmlTriggerCharacters = ['#', '.', '!', ',', '-', '<'];
+    private static readonly char[] s_razorTriggerCharacters = ['<', ':', ' '];
+    private static readonly char[] s_csharpTriggerCharacters = [' ', '(', '=', '#', '.', '<', '[', '{', '"', '/', ':', '~'];
+    private static readonly char[] s_commitCharacters = [' ', '>', ';', '='];
+
+    private readonly HashSet<char> _csharpTriggerCharacters;
+    private readonly HashSet<char> _delegationTriggerCharacters;
+    private readonly HashSet<char> _htmlTriggerCharacters;
+    private readonly HashSet<char> _razorTriggerCharacters;
+    private readonly ImmutableArray<string> _allTriggerCharacters;
+    private readonly ImmutableArray<string> _allCommitCharacters;
+
+    // Note: Create a new array each time to avoid modifying the original array.
```

Comment:
This is a scary comment. Do things actually modify this?

Reply:
If they did, it would be entirely out of our control and hard to find. Prior to this change, we just returning the same mutable array. If some JsonConverter or MessagePackFormatter in the LSP assumed that the incoming array was a fresh copy and modified it, that _could_ corrupt `CompletionTriggerAndCommitCharacters`. I learned to be fearful of such scenarios many years ago when we uncovered a bug in COM Interop that could corrupt the value of string.Empty.